plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -141,15 +147,18 @@ public class IcebergSplitSource
     private Iterator<FileScanTaskWithDomain> fileTasksIterator = emptyIterator();
 
     private final boolean recordScannedFiles;
+    private final int currentSpecId;
     private ImmutableSet.Builder<DataFileWithDeleteFiles> scannedFiles = ImmutableSet.builder();
+    @Nullable
+    private Map<StructLikeWrapperWithFieldIdToIndex, Optional<FileScanTaskWithDomain>> scannedFilesByPartition = new HashMap<>();
```

Comment:
I would argue that there is no need to log how many files were skipped.
Consider using a local variable in the method instead of adding more state to the class.

Reply:
This state is required regardless of the log line. We need to maintain `scannedFilesByPartition` for the entire lifetime of the split source as file tasks iteration is not guaranteed to be partition at a time.
Having the log line is useful for knowing what happened. We should probably extend OPTIMIZE to print some useful stats like this at the end of the execution.